### PR TITLE
Fix race condition bug in non-inplace ggml_compute_forward_diag_mask_f32

### PR DESCRIPTION
**Pull request by xaedes**
test-grad0.c catched a bug in ggml_compute_forward_diag_mask_f32 when running in threaded mode.

This only affects non-inplace occurences of diag_mask.

memcpy needs to be synchronized across threads to avoid race conditions, so just do it in INIT phase.

Use memcpy instead of the dup function, because the dup function just returns in INIT phase.

### CODE REVIEW

**Comment by ggerganov** [comment]
I already made a fix for this in https://github.com/ggerganov/llama.cpp/pull/1450

But will use your version as it is less hacky

**Comment by ggerganov** [comment]
Resolve the CI error + conflict and will merge it

**Comment by xaedes** [comment]
Oh didnt see that commit with the fixes. Trailing whitespace removed. Do you resolve the conflict or should I?

File: ggml.c [hunk 1]
```diff
@@ -10501,34 +10501,29 @@ static void ggml_compute_forward_diag_mask_f32(
     assert(src1->type == GGML_TYPE_I32);
     assert(ggml_nelements(src1) == 2);
 
+    const int ith = params->ith;
+    const int nth = params->nth;
+
     const int  n_past  =       ((int32_t *) src1->data)[0];
     const bool inplace = (bool)((int32_t *) src1->data)[1];
+
```
Comment (ggerganov) [inline ggml.c:10509]:
```suggestion

```

**Review by ggerganov — APPROVED** [review]

**Pull request merged**